File: modules/remotebackend/regression-tests/zeromq-backend.rb
#!/usr/bin/env ruby
require "rubygems"
require 'bundler/setup'
require 'json'
require 'zero_mq'
$:.unshift File.dirname(__FILE__)
require "backend"

h = Handler.new(Pathname.new(File.join(File.dirname(__FILE__),"remote.sqlite3")).realpath.to_s)

f = File.open "/tmp/remotebackend.txt.#{$$}","a"
f.sync = true

begin
  context = ZeroMQ::Context.new
  socket = context.socket ZMQ::REP
  socket.bind("ipc:///tmp/pdns.0") or raise "Cannot bind to IPC socket"

  while(true) do
    line = ""
    rc = socket.recv_string line
    # expect json
    input = {}
    line = line.strip
    f.puts "#{Time.now.to_f}: [zmq] #{line}"
    next if line.empty?
    begin
      input = JSON.parse(line)
      next unless input and input["method"]
      method = "do_#{input["method"].downcase}"
      args = input["parameters"] || []

      if h.respond_to?(method.to_sym) == false
         res = false
      elsif args.size > 0
         res, log = h.send(method,args)
      else
         res, log = h.send(method)
      end
      socket.send_string ({:result => res, :log => log}).to_json, 0
      f.puts "#{Time.now.to_f} [zmq]: #{({:result => res, :log => log}).to_json}"
    rescue JSON::ParserError
      socket.send_string ({:result => false, :log => "Cannot parse input #{line}"}).to_json
      f.puts "#{Time.now.to_f} [zmq]: #{({:result => false, :log => "Cannot parse input #{line}"}).to_json}"
      next
    end
  end
rescue SystemExit, Interrupt
end

File: modules/remotebackend/regression-tests/zeromq-backend.py
#!/usr/bin/env python

import zmq
import json
import os
from urllib.parse import parse_qs, urlparse
from backend import BackendHandler

def run(socket, handler):
    while True:
        message = socket.recv()
        try:
            message = json.loads(message.decode().strip())
            method = "do_%s" % message['method'].lower()
            args = message['parameters']
            handler.result = False
            handler.log = []
            if callable(getattr(handler, method, None)):
                getattr(handler, method)(**args)
                result = json.dumps({'result': handler.result,'log': handler.log})
                socket.send(result.encode())
        except KeyboardInterrupt as e3:
            return
        except BrokenPipeError as e2:
            raise e2
        except Exception as e:
            print(e)
            socket.send(json.dumps({'result':False}).encode())


def main():
    path = os.path.dirname(os.path.realpath(__file__))
    context = zmq.Context()
    socket = context.socket(zmq.REP)
    socket.bind("ipc:///tmp/pdns.0")
    handler = BackendHandler(options={'dbpath': os.path.join(path, 'remote.sqlite3')})

    try:
        run(socket, handler)
    except KeyboardInterrupt as e:
        pass

    os.unlink("/tmp/remotebackend.0")
 
main()
